benchmark-ganache.ts

```diff
@@ -0,0 +1,165 @@
+import * as assert from "assert";
+
+const chalk: any = require("chalk");
+import * as ethers from "ethers";
+import Ganache from "ganache";
+
+const url = "mainnet";
+const blockGasLimit = "0x2625A00"; // 40,000,000
+const blockNumber = 13700000;
+const defaultBalance = "0xffffffffffffffffffffff";
+const targetBalance = "0xffffffffffffffffffff";
+const convexAddress = "0xF403C135812408BFbE8713b5A23a04b3D48AAE31";
+const convexAbi = [
+  "function shutdownSystem() external",
+  "function owner() external view returns (address)"
+];
+
+main();
+
+async function main(): Promise<void> {
+  /*
+   * setup
+   */
+
+  console.log(chalk.bold("Setting up Ganache..."));
+  console.time("setup-ganache");
+  console.group();
+
+  const { ganache, provider } = await prepareGanache({
+    url,
+    blockNumber,
+    blockGasLimit,
+    defaultBalance,
+    deleteCache: false
+  });
+
+  const convex = new ethers.Contract(convexAddress, convexAbi, provider);
+  const ownerAddress = await convex.owner();
+
+  await fundAccounts({
+    provider,
+    accounts: [convexAddress, ownerAddress],
+    amount: targetBalance
+  });
+
+  const owner = await unlockAddress({
+    provider,
+    address: ownerAddress
+  });
+
+  console.groupEnd();
+  console.timeEnd("setup-ganache");
+  console.log("");
+
+  /*
+   * simulation
+   */
+
+  console.log(chalk.bold("Simulating shutdown..."));
+  console.time("simulate-shutdown");
+  console.group();
+  const tx = await convex.connect(owner).shutdownSystem({
+    gasLimit: "0x25317C0" // 39,000,000
+  });
+
+  const receipt = await provider.waitForTransaction(tx.hash);
+  console.groupEnd();
+  console.timeEnd("simulate-shutdown");
+  console.log("");
+
+  // @ts-ignore looks like ganche has a bad typing
+  await ganache.disconnect();
+}
+
+interface PrepareGanacheOptions {
+  url: string;
+  blockNumber: "latest" | number;
+  blockGasLimit: string;
+  defaultBalance: string;
+  deleteCache: boolean;
+}
+
+async function prepareGanache({
+  url,
+  blockNumber,
+  blockGasLimit,
+  defaultBalance,
+  deleteCache
+}: PrepareGanacheOptions): Promise<{
+  ganache: ReturnType<typeof Ganache.provider>;
+  provider: ethers.providers.JsonRpcProvider;
+}> {
+  const ganache = Ganache.provider({
+    fork: {
+      url,
+      blockNumber,
+      deleteCache
+    },
+    miner: { blockGasLimit },
+    wallet: {
+      // ganache expects value in ETH
+      defaultBalance: ethers.utils.formatEther(defaultBalance)
+    },
+    logging: {
+      quiet: true
+    },
+    legacyInstamine: true
```

Comment:
> Legacy instamining, where transactions are fully mined before the hash is returned, is deprecated and will be removed in the future.

Will this be disabled once v7 is out of beta? Wondering if it might make sense to switch this to false now just to future-proof the script a bit

Reply:
Nah, `legacyInstamine` will continue to exist in v7 stable. I have this here for two reasons:
- Less overhead between getting the txHash and getting the receipt
- Something's going on between Ganache and ethers and this script doesn't work if I turn this flag off!

(okay that first reason is just an excuse ;)
 
Planning to work with @davidmurdoch next week to figure out this and the various other bugs this exercise has helped us find :)

Reply:
IIRC, Ethers v5 doesn't subscribe to newHeads events when the provider is an EIP-1193 provider, so it falls back to polling evert 4 seconds. As to why it doesn't work at all? I'm not sure and will have to dig in further. 

Ethers v6 will support the subscription methods though (@ricmoo please correct me if I'm wrong here 😁)

Reply:
You are correct; I wrote the subscriber classes yesterday. ;)